pkg/apps/strategy/support/lifecycle.go

```diff
@@ -224,58 +226,101 @@ func (e *hookExecutor) executeExecNewPod(hook *appsv1.LifecycleHook, rc *corev1.
 		fmt.Fprintf(e.out, "--> %s: Running hook pod ...\n", label)
 	}
 
-	stopChannel := make(chan struct{})
-	defer close(stopChannel)
-	nextPod := newPodWatch(e.pods.Pods(pod.Namespace), pod.Name, stopChannel)
-
-	// Wait for the hook pod to reach a terminal phase. Start reading logs as
-	// soon as the pod enters a usable phase.
 	var updatedPod *corev1.Pod
-	wg := &sync.WaitGroup{}
-	wg.Add(1)
 	restarts := int32(0)
 	alreadyRead := false
-waitLoop:
-	for {
-		updatedPod = nextPod()
-		switch updatedPod.Status.Phase {
-		case corev1.PodRunning:
-			completed = false
-
-			// We should read only the first time or in any container restart when we want to retry.
-			canRetry, restartCount := canRetryReading(updatedPod, restarts)
-			if alreadyRead && !canRetry {
-				break
-			}
-			// The hook container has restarted; we need to notify that we are retrying in the logs.
-			// TODO: Maybe log the container id
-			if restarts != restartCount {
-				wg.Add(1)
-				restarts = restartCount
-				fmt.Fprintf(e.out, "--> %s: Retrying hook pod (retry #%d)\n", label, restartCount)
+	wg := &sync.WaitGroup{}
+	wg.Add(1)
+
+	listWatcher := &cache.ListWatch{
+		ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+			options.FieldSelector = fields.OneTermEqualSelector("metadata.name", pod.Name).String()
+			return e.pods.Pods(pod.Namespace).List(options)
+		},
+		WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+			options.FieldSelector = fields.OneTermEqualSelector("metadata.name", pod.Name).String()
+			return e.pods.Pods(pod.Namespace).Watch(options)
+		},
+	}
+	// make sure that the pod exists and wasn't deleted early
+	preconditionFunc := func(store cache.Store) (bool, error) {
+		_, exists, err := store.Get(&metav1.ObjectMeta{Namespace: pod.Namespace, Name: pod.Name})
+		if err != nil {
+			return true, err
+		}
+		if !exists {
```

Comment:
you could also verify uid, to avoid case of re-create, but that would be super rare